### DISCUSSION TITLE
API

### DISCUSSION THREAD

**Discussion by mcguirepr89** [mcguirepr89/BirdNET-Pi#310] [Show and tell]
I'm building this -- here is the POC
https://virginia.birdnetpi.com/api/detections

**Comment by ehpersonal38** [comment]
Looks good! What do you think the intended use case of the API is? People just wanting to make their own interface of sorts, or more for research groups to extract data to spreadsheets and such?

**Reply by mcguirepr89** [reply]
Just interfacing for other folks mostly

**Reply by MyosQ** [reply]
I would be very interested in this feature. For building a custom interface and for building an interface with overview over many devices. If I understand it correctly, now you can only view one device at a time in the web interface?

**Reply by mcguirepr89** [reply]
Thanks, @MyosQ for weighing in on adding the API. I would really like it if folks like yourself were to also weigh in on which interfaces (frameworks ?) you would be interested in using -- it can help me both with configuring the API and in preparing the repo for the new model to be built from scratch using the framework for which I've been presented the most compelling use case

**Reply by MyosQ** [reply]
My first thought was to do a dashboard of some sort. One single page that updates its content automatically. In it could be for example: species today, top 3 species this week and new species this week. You could also have graphs and other visualisations. 

R shiny is quite handy for dashboards and would be my first choice. But I think react could also work.

**Comment by DaveBKC** [comment]
Would this allow users to grab data from the BirdNET-Pi local database? I'm interested in having a copy of that BirdNet data in my  mariadb weather data database.  From there I would import into Tableau.

**Reply by mcguirepr89** [reply]
Yes -- the POC in the `api` branch is not production ready, and only currently has one end-point configured which returns the entire database in JSON format. I need to figure the best server to use for production since I'm only just now getting familiar with Flask (Caddy will stand in front as the reverse_proxy).

**Reply by mcguirepr89** [reply]
It's also not active right now (I'm on the `main` branch currently)

**Comment by timsterc** [comment]
If this isn't the right place to mention this - please feel free to delete - or suggest where I might better share.

The goal with BirdWeather is to provide both live visualization, but also to be a fully open databank of all the detections uploaded.  We're using GraphQL - so for those familiar with it:
- API - https://app.birdweather.com/api/index.html
- Online GraphQL Query Tool - https://app.birdweather.com/graphiql

We've also finished the first pass at our data explorer - https://app.birdweather.com/data - which also supports exporting results via CSV.

Cheers!
-Tim

**Reply by becausebirds** [reply]
Your API for BirdWeather is great! I'd be interested in requesting detections only for the current day, instead of having the 24-hour rolling time period only. I built a digital counter and am using it to display the detection count and latest species detected. It would be amazing to see just the day's data on it!

**Comment by London-Lass** [comment]
I'm involved with a livestream of wildlife in Scotland using OBS (Birds, Mammals etc) and think that integrating with BirdNET would be a a fantastic feature. We already allow users to spot birds etc, with an index being created, and videos created with thumbnails of species.  (For anyone interested, search Youtube for "Scottish Wildlife Garden" if you want to see the livestream,)

What is the current status of the API available for BirdNET-Pi so that I might on say query over a period of time which species have been heard, so that this can then be shared with viewers? The initial comment on this discussion suggests that it just returns the entire database in json format. Any progress since then?

**Reply by ehpersonal38** [reply]
This looks like a cool project! I've seen a few livestreams like this on YouTube before and think they're pretty neat. 

The current status of the API for BirdNET-Pi is that it's non-existent. However, I think it'd be easy enough to code up something that would meet your use-case. So, let me know what format you need (json, csv, something else?) and I'll try to get some type of rudimentary API worked up. Also do you currently have a BirdNET-Pi installation up and running?

**Comment by London-Lass** [comment]
Thanks for getting back so quickly. From an api perspective, a json response is perfect. Ideally I'd be able to pass parameters to filter on say last hour, day, week or whatever so that I can produce a video of "Birds heard in the last XXX" with thumbnails etc.

The heart of the livestream is a bunch of flask microservices to manage interaction with youtube chat, media creation, sequencing of view changes according to time of the day at the place etc. It should be an easy task to interact with your api and do something with the resulting data.

I have a pi or two in a drawer at home, and possibly a USB sound card, but I think i'll need to make some investments before I'd be able to listen in with any confidence.

**Reply by ehpersonal38** [reply]
> Ideally I'd be able to pass parameters to filter on say last hour, day, week or whatever so that I can produce a video of "Birds heard in the last XXX" with thumbnails etc.

This is certainly possible! So you would only need the JSON data and no bird images? I could include Flickr images for each detection, if you need that. But if you have your own images then that's great too.

> I have a pi or two in a drawer at home, and possibly a USB sound card, but I think i'll need to make some investments before I'd be able to listen in with any confidence.

Do let me know when you have your system up and running - then I'll start building the API. 🙂 

I see you're in the UK, so if you want an amazing quality mic here is the one that's highly recommended - https://micbooster.com/clippy-and-pluggy-microphones/65-clippy-em172-microphone.html. You'll also need a TRS sound card for it, like: https://www.amazon.com/UGREEN-External-Headphone-Microphone-Desktops/dp/B01N905VOY/?pldnSite=1&th=1

You can also build your own mic with the same high quality audio for about 10 euros, you'll just need a few hours and some soldering skills - https://github.com/mcguirepr89/BirdNET-Pi/discussions/39#discussioncomment-2180372

Either way, best of luck ~~

**Reply by London-Lass** [reply]
Managed to get it running last night using what I had. Install went smoothly, and despite a strong whine from the Mic and Soundcard it managed to identify some random bird calls being played through youtube. 

Sound card ordered, will investigate buying or making a suitable mic for purposes next.

**Reply by becausebirds** [reply]
I currently have a digital counter setup (enclosure coming soon) that shows my station detection count from BirdWeather. It works great but it would be nice to use an API on the BirdNET-Pi itself. 

My main desire for the API is for one of the time period parameter options to be "today". Ideally, my counter will display the detection count for the current day only.

<img width="775" alt="Screenshot 2023-02-12 at 11 31 32 AM" src="https://user-images.githubusercontent.com/110354170/218327001-69fa8e05-a313-4050-93de-a3c16e743c5e.png">

**Reply by ehpersonal38** [reply]
Hi @becausebirds 

This is really easy to implement if you want to do on your own:
1. Create a new file `/home/pi/BirdNET-Pi/scripts/today_count.php`
2. Copy the following into the file:
```php
<?php
$db = new SQLite3('./scripts/birds.db', SQLITE3_OPEN_CREATE | SQLITE3_OPEN_READWRITE);
if($db == False){
  echo "Database is busy";
  header("refresh: 0;");
}

$statement = $db->prepare('SELECT COUNT(*) FROM detections WHERE Date == DATE(\'now\', \'localtime\')');
if($statement == False){
  echo "Database is busy";
  header("refresh: 0;");
}
$result = $statement->execute();
$todaycount = $result->fetchArray(SQLITE3_ASSOC);

echo $todaycount['COUNT(*)'];
```
3. Symbolically link the file to the web root: `ln -fs ~/BirdNET-Pi/scripts/today_count.php ~/BirdSongs/Extracted/`

Now whenever you access `/today_count.php` on your BirdNET-Pi, it will display the number of detections it's had today. 

You can JSONify this if you need as well - replace `echo $todaycount['COUNT(*)'];` with `echo json_encode( array("today_count" => $todaycount['COUNT(*)']) );`

**Reply by becausebirds** [reply]
Thank you for this! I've created the file with the code modification you mentioned to output JSON. 

When I visit [https://birdnet.becausebirds.com/today_count.php](https://birdnet.becausebirds.com/today_count.php) it just shows the contents of the file. I'm new at a lot of this, so bear with me while I'm learning :D. For my application I've been making a GET request to the BirdWeather api URL. Do I do the same thing but to the above URL? 

I tried but the contents returned are the same as what I see in my browser (the raw code in the PHP file).

Thank you!!!

**Reply by ehpersonal38** [reply]
> I tried but the contents returned are the same as what I see in my browser (the raw code in the PHP file).

I think this is a problem with the text editor you used to paste the contents into the file with. To make it easy just run this command and it'll download the proper version of `today_count.php` for you and put it in the correct directory:
```
wget https://gist.githubusercontent.com/ehpersonal38/a34ce000e1645bc85aba936a5f89ac72/raw/43311739c3176f1b77e298ae5fabd58afc27250b/today_count.php -O ~/BirdNET-Pi/scripts/today_count.php
```

> For my application I've been making a GET request to the BirdWeather api URL. Do I do the same thing but to the above URL?

Yes it should be the same! You'll just have to edit the JSON key that your application looks for.

**Reply by becausebirds** [reply]
THANK YOU! Works like a charm. My sign is up and running with today's count only. So happy.

**Reply by ehpersonal38** [reply]
@becausebirds Hey I saw your Reddit post about this! Super cool to see the reach of this project, I bet Patrick would be really happy to see his work positively affecting so many people. And thank you for taking the time to share it.

**Reply by becausebirds** [reply]
I'm so happy you saw it! Reddit received it so wonderfully. So many curious and inspired people picking up the project. You were instrumental in getting me over the finish line!

BirdNET-Pi deserves as much attention as possible! I'm thrilled to have been able to create such as fun project from it! I look forward to a local API!
